sql/src/main/java/io/deephaven/sql/SqlRootContext.java

```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.sql;
+
+import org.apache.calcite.rel.RelNode;
+
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+
+final class SqlRootContext {
+
+    public static SqlRootContext of(RelNode root, Scope scope) {
+        return new SqlRootContext(root, encounterOrder(root), scope);
+    }
+
+    private final RelNode root;
+    private final Map<RelNode, Integer> repeatableId;
+    private final Scope scope;
+
+    private SqlRootContext(RelNode root, Map<RelNode, Integer> repeatableId, Scope scope) {
+        this.root = Objects.requireNonNull(root);
+        this.repeatableId = Objects.requireNonNull(repeatableId);
+        this.scope = Objects.requireNonNull(scope);
+    }
+
+    public RelNode root() {
+        return root;
+    }
+
+    public Scope scope() {
+        return scope;
+    }
+
+    public NamedAdapter namedAdapter() {
+        return new NamedAdapter(this);
+    }
+
+    public IndexRef createIndexRef(Prefix context, RelNode node) {
+        // Calcite RelNode#getId is distinct for every relational expression that is created, even if the same SQL query
+        // string is parsed again. From a unit testing perspective, and a server cacheability perspective, we would
+        // prefer to assign consistent (internal) column names for the same SQL query string.
+        final String columnPrefix = String.format("%s%d_", context.prefix(), repeatableId(node));
+        return new IndexRef(this, columnPrefix, 0);
+    }
+
+    private int repeatableId(RelNode node) {
+        final Integer id = repeatableId.get(node);
+        if (id == null) {
+            throw new IllegalArgumentException("node does not belong to this context");
```

Comment:
This should never happen, but it's cryptic. Should be clearer, and include more info (node, repeatable ID map, etc).